src/dbnode/storage/series/buffer.go

```diff
@@ -89,53 +99,50 @@ type databaseBuffer interface {
 
 	Stats() bufferStats
 
-	// MinMax returns the minimum and maximum blockstarts for the buckets
-	// that are contained within the buffer. These ranges exclude buckets
-	// that have already been drained (as those buckets are no longer in use.)
-	MinMax() (time.Time, time.Time, error)
-
-	Tick() bufferTickResult
-
-	NeedsDrain() bool
+	Tick(versions map[xtime.UnixNano]BlockState) bufferTickResult
 
-	DrainAndReset() drainAndResetResult
-
-	Bootstrap(bl block.DatabaseBlock) error
+	Bootstrap(bl block.DatabaseBlock)
 
 	Reset(opts Options)
 }
 
 type bufferStats struct {
-	openBlocks  int
 	wiredBlocks int
 }
 
-type drainAndResetResult struct {
-	mergedOutOfOrderBlocks int
-}
-
 type bufferTickResult struct {
 	mergedOutOfOrderBlocks int
+	evictedBuckets         int
 }
 
 type dbBuffer struct {
-	opts              Options
-	nowFn             clock.NowFn
-	drainFn           databaseBufferDrainFn
-	pastMostBucketIdx int
-	buckets           [bucketsLen]dbBufferBucket
-	blockSize         time.Duration
-	bufferPast        time.Duration
-	bufferFuture      time.Duration
+	opts  Options
+	nowFn clock.NowFn
+
+	bucketsMap map[xtime.UnixNano]*BufferBucketVersions
+	// Cache of buckets to avoid map lookup of above.
+	bucketVersionsCache [bucketsCacheSize]*BufferBucketVersions
+	// This is an in order slice of the block starts in the bucketsMap.
+	// We maintain this to avoid sorting the map keys adhoc when we want to
+	// perform operations in chronological order.
+	inOrderBlockStarts []time.Time
+	bucketVersionsPool *BufferBucketVersionsPool
+	bucketPool         *BufferBucketPool
+
+	blockSize             time.Duration
```

Comment:
should we replace term "block" with "bucket" everywhere inside Buffer?

Reply:
Hmm I think the term block fits more in certain areas in the buffer. While buckets map to blocks, they don't really logically have a size to them, only a reference to the block start that it corresponds to. Buckets refer to a bucket of encoders, but the buffer also deals with blocks (of data), which is a different concept.